### ISSUE TITLE
Enable client-side resizing when uploading an image

### ISSUE THREAD

**Issue by asaunier** [c2corg/v6_ui#1046]
There is an option in ng-file-upload, ``ngf-resize``. See https://github.com/c2corg/v6_ui/issues/929#issuecomment-262923465

**Comment by asaunier** [comment]
ng-file-upload config is here: https://github.com/c2corg/v6_ui/blob/master/c2corg_ui/static/partials/imageuploader.html#L1-L12

> And what would be considered as a "too big image" and what would be the "correct, smaller size"?
> We can take V5 values : size < 8192 x 2048 px AND weight < 2 Mo

It is possible to specify for what dimensions an image should be resized:
```
ngf-resize-if="$width > 1000 || $height > 1000" or "resizeCondition($file, $width, $height)"
```
I don't know if it is possible to test the image weight? Perhaps using ``$file``?

At least it is possible to add limitations/validations and refuse images that would be too big (or too small...). See the validations section in the reference at https://github.com/danialfarid/ng-file-upload#file-select-and-drop
For instance ``ngf-min-size, ngf-max-size, ngf-max-total-size="100" in bytes or "'10KB'" or "'10MB'" or "'10GB'"``  or ``ngf-min-height, ngf-max-height, ngf-min-width, ngf-max-width="1000" in pixels``.

Possible resizing options:
```
ngf-resize="{width: 100, height: 100, quality: .8, type: 'image/jpeg', 
               ratio: '1:2', centerCrop: true, pattern='.jpg', restoreExif: false}" 
               or resizeOptions() // a function returning a promise which resolves into the options.
    // resizes the image to the given width/height or ratio. Quality is optional between 0.1 and 1.0), 
    // type is optional convert it to the given image type format.
    // centerCrop true will center crop the image if it does not fit within the given width/height or ratio. 
    // centerCrop false (default) will not crop the image and will fit it within the given width/height or ratio 
    // so the resulting image width (or height) could be less than given width (or height).
    // pattern is to resize only the files that their name or type matches the pattern similar to ngf-pattern.
    // restoreExif boolean default true, will restore exif info on the resized image.
```

See the demo at https://angular-file-upload.appspot.com/

**Comment by ginold** [comment]
Exactly what i was looking for! I will start to work on this today.

**Comment by tsauerwein** [comment]
Please test if the resizing doesn't freeze the browser when trying to add a lot of images at once (especially on slow machines).

**Comment by asaunier** [comment]
@ginold please note it is not a high prio task (tagged 6.1 - personally I think it would help solve the image upload problem so I would have kept it 6.0 :P )

Rather focus on the associations removal PR and on the cards displaying/sorting https://github.com/c2corg/v6_ui/issues/690

**Comment by ginold** [comment]
The issue that this one is created from was tagged as blocking. Has it changed priority?

**Comment by stef74** [comment]
Upload image is blocking ... but we unblock by uncrease timeout server. So this issue is important but not blocking.

**Comment by ginold** [comment]
Ok, I didn't know we solved the timeout issue :+1:

**Comment by asaunier** [comment]
> Ok, I didn't know we solved the timeout issue 

It's not! (yet)

The initial issue has been split in 2 tasks: only the timeout setting remains in https://github.com/c2corg/v6_ui/issues/929#issuecomment-263864176

Concerning the current issue (ng resize), as you can see it is not tagged blocking and its milestone is 6.1.

I have the feeling it could be not that complicated. I have tried yesterday to add ``ngf-resize="{width: 300, height: 300}"``=> it worked out of the box. But something more sophisticated is needed, with validation (and error messages if the validation fails?!), resizing conditions, and of course testing that the results are acceptable in terms of quality (+ the freezing issue suspected by @tsauerwein )

**Comment by asaunier** [comment]
Moving to milestone golive because I think it would really help solving https://github.com/c2corg/v6_ui/issues/929
(but not high prio)

**Comment by asaunier** [comment]
First version of client resizing here: #1113

**Comment by stef74** [comment]
I have try to upload 10 images ... half have timeout

**Comment by stef74** [comment]
But no image at all in WP https://www.demov6.camptocamp.org/routes/828282/fr/jo-part-a-cetamol

**Comment by mgerbaux** [comment]
client resizing seems not to work for me… (6Mo and 8Mo  images, blocked at 14% uploading)

**Comment by loicperrin** [comment]
Uploading KO with the 2 images I have tried (2.7 and 2.6 Mo)

**Comment by asaunier** [comment]
The resizing dimensions and quality are probably still too big:
``{width: 4096, height: 2048, quality: 1.0}``

https://github.com/c2corg/v6_ui/blob/306abd753996259939e02599cc0f5d537d5ef256/c2corg_ui/static/partials/imageuploader.html#L12
https://github.com/c2corg/v6_ui/blob/master/c2corg_ui/static/js/imageuploader.js#L400-L402

I would suggest ``{width: 2048, height: 2048, quality: 0.9}``.
OK for you?

**Comment by asaunier** [comment]
I have made a new PR #1168

2 changes:
* images are resized to max 2048px and with quality 90% instead of 100% => hopefully it will reduce the weight of images. But I have not been able to test yet, I still have to find big enough images :P 
* resizing is triggered only if the image is bigger the 2 MB, I have dropped the width/height constraint. This way a panorama of 10000 x 200px will be uploaded as is if it is < 2 MB.

Sounds good?

**Comment by mgerbaux** [comment]
> Sounds good?

yes for me

**Comment by asaunier** [comment]
OK I have just tested my PR with images from 4 to 20 MB, the results are pretty good.
One improvement that could be added: until the image is actually resized client-side, nothing is shown in the upload interface, one can think nothing happens and upload the image again (but it's still pretty fast on my PC with the 20MB image).
It would be useful to show the image upload bar (with no overview until it is ready) when the resize process has started, not only when the upload actually begins.
